### PR TITLE
Correctly extract username/password from URL

### PR DESCRIPTION
**Pull request by gregatcode**
Corrected regexp. Otherwise it was impossible to use urls with credentials **and** @ character somewhere in the path url segment - it was capturing the whole part of the url (down to the last @ char) as a password.

Not sure if i need to add unit test for this small edge case.

### CODE REVIEW

File: src/Composer/Util/RemoteFilesystem.php [hunk 1]
```diff
@@ -207,7 +207,7 @@ protected function get($originUrl, $fileUrl, $additionalOptions = array(), $file
         $this->redirects = 1; // The first request counts.
 
         // capture username/password from URL if there is one
-        if (preg_match('{^https?://(.+):(.+)@([^/]+)}i', $fileUrl, $match)) {
+        if (preg_match('{^https?://([^:]+):([^@]+)@([^/]+)}i', $fileUrl, $match)) {
```
Comment (stof) [inline src/Composer/Util/RemoteFilesystem.php]: it must also forbid `/` in both of them (otherwise a URL with a `:` and a `@` in the query string could be matched as authenticated)

**Comment by stof** [comment]
and yes, you should add a unit test. Such edge cases are precisely cases for which a unit test is the only way to avoid breaking them again in the future (it took years to identify this bug, and it could take years again to catch a regression)

File: src/Composer/Util/RemoteFilesystem.php (hunk 1 above)
Comment (gregatcode) [reply src/Composer/Util/RemoteFilesystem.php]: Exactly. Probably, parse_url() would be more feasible here.

**Comment by Seldaek** [comment]
Thanks, merged in 1.5 and master
